<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>Asynchronous JavaScript</title>
</head>
<body>
    <h1>Asynchronous JavaScript</h1>
    <script>
        /* 1st example
        const second = () => {
            // 1st argument is action of the function
            // 2nd argument is the time out of the funcitno in milliseconds. 
            setTimeout(() => {
                console.log('Async Hey There'); 
            }, 2000);
        }

        const first = () => {
            console.log('Hey There');
            // after 2s, see second(), see the 'The end' first
            // This is called aSynchronous
            // Code run line by line called Synchrounous
            // purpose of aSync is to have multiple process run at the same time
            // See notes for event loop, execution stack and Web API
            second(); 
            console.log('The end'); 
        }
        first(); 
        */

        /* Example of CallBack Hell 
        Promises will solve this issue
        function getRecipe(){
            setTimeout(() => {
                const recipeID = [134,4531,4343,342]; 
                console.log(recipeID);

                // 3rd argument is the argument will be passed into the function
                setTimeout((id) => {
                    const recipe = {
                        title: 'Fresh Tomato pasta', 
                        publisher: 'Trung'
                    }; 
                    console.log(`${id}: ${recipe.title}`); 

                    setTimeout((publisher) => {
                        const recipe2 = {
                            title: 'Italian Pizza', 
                            publisher: 'Nigga Please'
                        }; 

                        console.log(recipe2); 
                    }, 1500, recipe.publisher);
                }, 1500, recipeID[2]);
            }, 1000);
        }
        getRecipe(); 
        */

        // PROMISE IS AN OBJECT THAT 
        // - KEEPS TRACK ABOUT WHETHER A CERTAIN EVENT HAS HAPPENED ALREADY OR NOT
        // - DETERMINE WHAT HAPPENS AFTER THE EVENT HAS HAPPENED
        // - IMPLEMENTS THE CONCEPT OF A FUTURE VALUE TAHT WE'RE EXPECTING

        // PROMISE STATE: 
        // - PENDING: THE EVENT HASN'T HAPPENED
        // - SETTLED / RESOLVED: THE EVENT HAS HAPPENED: 
        //   +, SUCCESSFUL: IF EVENT SUCCESFULLY RUN 
        //   +, REJECTED: IF EVENT UNSUCCESSFULLY RUN  

        // Making example above using promise
        // passing 2 states into the function inside promise
        const getIDs = new Promise((resolve, reject) => {
            setTimeout(() => {
                // successful then return the array
                resolve([134,4531,4343,342]);
                // Promise is always successful so you'll have to 
                // comment out the resolve function to see the output of reject()
                reject('There is an error dawg !!');
                // Run the reject() function to return 
                // if the function is not successful
            }, 1500);
        });

        const getRecipe = recID => {
            return new Promise((resolve, reject) =>{

                setTimeout((ID) => {
                    const recipe = {
                        title: 'Fresh Tomato pasta', 
                        publisher: 'Trung'
                    }; 
                    resolve(`${ID}: ${recipe.title}`); 
                }, 1500, recID);
            }); 
        };

        const getRelated = (publisher) => {
            return new Promise((resolve, reject) => {
                setTimeout((pub) => {
                    const recipe2 = {
                            title: 'Italian Pizza', 
                            publisher: 'Nigga Please'
                        }; 

                    resolve(`${pub}: ${recipe2.title}`); 
                }, 1500, publisher);
            });
        };

        // then() add event handler in case the event is resolved
        // IDs will be the result of the resolve function
        // in this case it will be the entire array
        getIDs
        .then((IDs) => {
            // Take the return of resolve() Function above
            console.log(IDs); 
            return getRecipe(IDs[2]); 
        })
        .then(recipe => {
            console.log(recipe); 
            return getRelated('Trung'); 
        })
        .then(recipe => {
            console.log(recipe); 
        })
        .catch(error => {
            // Take the return of reject() Function above
            console.log(error); 
        });
    </script>
</body>
</html>